agent/yubiagent/slot.go

```diff
@@ -0,0 +1,191 @@
+	attestationCert, err := yubiAgent.AttestSlot(code)
+	if err != nil {
+		return nil, err
+	}
+	// Due to Infineon Technologies' RSA key generation issue, we do not support RSA
+	// certificates generated by YubiKeys with firmware between 4.2.6-4.3.4.
+	// Ref: https://www.yubico.com/support/security-advisories/ysa-2017-01/
+	if attestationCert.PublicKeyAlgorithm == x509.RSA && isAffectedVersion(attestationCert.Extensions) {
+		return nil, fmt.Errorf("found RSA certificate generated by YubiKey with firmware 4.2.6-4.3.4")
+	}
+
+	publicKey, err := ssh.NewPublicKey(attestationCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	if validateSSHPublicKeyAlgo(publicKey) {
+		return nil, fmt.Errorf("unsupported certificate type in the key slot: %s", code)
+	}
+
+	touchPolicy := getTouchPolicy(attestationCert)
+
+	return &SlotAgent{
+		code:      code,
+		public:    publicKey,
+		attest:    attestationCert,
+		policy:    touchPolicy,
+		yubiAgent: yubiAgent,
+	}, nil
+}
+
+// NewSlotAgentWithAttrs returns a new slot agent with attribute values.
+func NewSlotAgentWithAttrs(yubiAgent YubiAgent, code string, public ssh.PublicKey,
+	attest *x509.Certificate, policy keyid.TouchPolicy) *SlotAgent {
+	return &SlotAgent{
+		yubiAgent: yubiAgent,
+		code:      code,
+		public:    public,
+		attest:    attest,
+		policy:    policy,
+	}
+}
+
+// RegisterCSR registers the given certificate signing request for the key slot.
+func (s *SlotAgent) RegisterCSR(csr *proto.SSHCertificateSigningRequest) {
+	s.csrs = append(s.csrs, csr)
+}
+
+// RegisterCSRs registers multiple certificate signing request for the key slot.
+func (s *SlotAgent) RegisterCSRs(csrs []*proto.SSHCertificateSigningRequest) {
+	for _, csr := range csrs {
+		s.RegisterCSR(csr)
+	}
+}
+
+// CSRs returns the registered CSRs.
+func (s *SlotAgent) CSRs() []*proto.SSHCertificateSigningRequest {
+	return s.csrs
+}
+
+// AddCertsToAgent adds certificates to the agent.
+func (s *SlotAgent) AddCertsToAgent(certs []ssh.PublicKey, comments []string) error {
+	for i, c := range certs {
+		cert, err := keyutil.CastSSHPublicKeyToCertificate(c)
+		if err != nil {
+			continue
+		}
+		err = s.yubiAgent.AddHardCert(cert, comments[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PublicKey returns the public key of the key slot.
+func (s *SlotAgent) PublicKey() ssh.PublicKey {
+	return s.public
+}
+
+// TouchPolicy returns the touch policy of the key slot.
+func (s *SlotAgent) TouchPolicy() keyid.TouchPolicy {
+	return s.policy
+}
+
+// Serial returns the serial number of the yubikey.
+func (s *SlotAgent) Serial() (string, error) {
+	return yubiattest.ModHex(s.attest)
+}
+
+// SlotCode returns the code number of the key slot.
+func (s *SlotAgent) SlotCode() string {
+	return s.code
+}
+
+// AttestCert returns the attest cert of the key slot.
+func (s *SlotAgent) AttestCert() *x509.Certificate {
+	return s.attest
+}
+
+// Agent returns the yubiagent.
+func (s *SlotAgent) Agent() YubiAgent {
```

Comment:
Why we want to export this? Can you provide use cases?

Reply:
There are 2 use cases:
1) During ssh key challenge, we will need the agent of that key slot to invoke "Sign()"
https://github.com/theparanoids/ysshra/blob/main/agent/ssh/agent.go#L74-L85
2) During attestation, I want to use the same agent of that key slot to fetch the f9 certificate.
```
f9Cert, err := a.slot.Agent().ReadSlot("f9")
```
For (2), it can be eliminated by creating a slot object for f9. For (1), I don't like to expose another function `Sign()` for `struct Slot`.

Reply:
This doesn't sound right. It make more sense to implement `Sign()` for a `Slot` and for the f9, it should be hold inside the struct
```
type authn struct {
	attest *yubiagnet.Slot
	slotAgent *yubiagent.Slot
	conf      *conf
}
```